// tracerbench/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Report = exports.resolveTitles = exports.generateDataForHTML = exports.Compare = exports.Profile = exports.CompareReport = exports.RecordHAR = void 0;
const compare_1 = require("./commands/compare");
exports.Compare = compare_1.default;
const report_1 = require("./commands/compare/report");
exports.CompareReport = report_1.default;
exports.Report = report_1.default;
const profile_1 = require("./commands/profile");
exports.Profile = profile_1.default;
const record_har_1 = require("./commands/record-har");
exports.RecordHAR = record_har_1.default;
const create_consumable_html_1 = require("./helpers/create-consumable-html");
Object.defineProperty(exports, "generateDataForHTML", { enumerable: true, get: function () { return create_consumable_html_1.generateDataForHTML; } });
Object.defineProperty(exports, "resolveTitles", { enumerable: true, get: function () { return create_consumable_html_1.resolveTitles; } });
var command_1 = require("@oclif/command");
Object.defineProperty(exports, "run", { enumerable: true, get: function () { return command_1.run; } });
__exportStar(require("./helpers"), exports);
