// tracerbench@8.0.1/dist/src/helpers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logHeading = exports.chalkScheme = exports.md5sum = exports.parseMarkers = exports.getEmulateDeviceSettingForKeyAndOrientation = exports.deviceSettings = void 0;
const device_settings_1 = require("./device-settings");
exports.deviceSettings = device_settings_1.default;
Object.defineProperty(exports, "getEmulateDeviceSettingForKeyAndOrientation", { enumerable: true, get: function () { return device_settings_1.getEmulateDeviceSettingForKeyAndOrientation; } });
const utils_1 = require("./utils");
Object.defineProperty(exports, "parseMarkers", { enumerable: true, get: function () { return utils_1.parseMarkers; } });
Object.defineProperty(exports, "md5sum", { enumerable: true, get: function () { return utils_1.md5sum; } });
Object.defineProperty(exports, "chalkScheme", { enumerable: true, get: function () { return utils_1.chalkScheme; } });
Object.defineProperty(exports, "logHeading", { enumerable: true, get: function () { return utils_1.logHeading; } });
