// @tracerbench/core/dist/trace/archive-trace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleHeaders = exports.processEntries = exports.recordHARClient = void 0;
const utils_1 = require("./utils");
async function recordHARClient(url, cookies, marker, conditions, altBrowserArgs) {
    const networkRequests = [];
    const archive = {
        log: {
            version: '0.0.0',
            creator: {
                name: 'TracerBench',
                version: '0.0.0'
            },
            entries: []
        }
    };
    const browserArgs = utils_1.getBrowserArgs(altBrowserArgs);
    const browser = await utils_1.createBrowser(browserArgs);
    try {
        const chrome = await utils_1.getTab(browser.connection);
        chrome.on('Network.requestWillBeSent', (params) => {
            console.log(`RECORDING-REQUEST :: ${params.request.method} :: ${params.type} :: ${params.request.url}`);
        });
        chrome.on('Network.responseReceived', (params) => {
            networkRequests.push(params);
        });
        // enable Network / Page / Runtime
        await Promise.all([
            chrome.send('Page.enable'),
            chrome.send('Network.enable'),
            chrome.send('Runtime.enable')
        ]); // clear and disable cache
        await chrome.send('Network.clearBrowserCache');
        // disable cache
        await chrome.send('Network.setCacheDisabled', { cacheDisabled: true });
        await utils_1.emulate(chrome, conditions);
        // set cookies
        await utils_1.setCookies(chrome, cookies);
        // add performance observer script to eval
        await chrome.send('Page.addScriptToEvaluateOnNewDocument', {
            source: `
        self.__TBMarkerPromise = new Promise(resolve => {
          const observer = new PerformanceObserver((list) => {
            if (list.getEntriesByName('${marker}').length > 0) {
              resolve();
            }
        });
        observer.observe({ entryTypes: ["mark", "navigation"] });
        });`
        });
        // navigate to the url
        await chrome.send('Page.navigate', { url });
        // eval
        const evalPromise = chrome.send('Runtime.evaluate', {
            expression: `__TBMarkerPromise`,
            awaitPromise: true
        });
        let timeoutId;
        const timeout = new Promise((reject) => {
            timeoutId = setTimeout(() => {
                clearTimeout(timeoutId);
                reject('Promise timed out after waiting for 20 seconds');
            }, 20000);
        });
        await Promise.race([evalPromise, timeout]).then(() => {
            if (timeoutId) {
                clearTimeout(timeoutId);
            }
        });
        archive.log.entries = await processEntries(networkRequests, chrome);
        await Promise.all([
            chrome.send('Network.disable'),
            chrome.send('Runtime.disable')
        ]);
        await chrome.send('Page.close');
    }
    catch (e) {
        throw new Error(`Network Request could not be captured. ${e}`);
    }
    finally {
        if (browser) {
            await browser.dispose();
        }
    }
    return archive;
}
exports.recordHARClient = recordHARClient;
async function processEntries(networkRequests, chrome) {
    const entries = [];
    for (let i = 0; i < networkRequests.length; i++) {
        console.log(`BUILDING-HAR-ENTRY-FOR :: ${networkRequests[i].response.url}`);
        const requestId = networkRequests[i].requestId;
        const response = networkRequests[i].response;
        const { body } = await chrome.send('Network.getResponseBody', {
            requestId
        });
        const { url, requestHeaders, status, statusText, headers } = response;
        const entry = {
            request: {
                url,
                method: '',
                httpVersion: '',
                cookies: [],
                headers: handleHeaders(requestHeaders),
                queryString: [],
                headersSize: 0,
                bodySize: 0
            },
            response: {
                status,
                statusText,
                httpVersion: '',
                cookies: [],
                headers: handleHeaders(headers),
                redirectURL: '',
                headersSize: 0,
                bodySize: 0,
                content: {
                    text: body,
                    size: 0,
                    mimeType: ''
                }
            },
            time: 0,
            cache: {},
            timings: {
                send: 0,
                wait: 0,
                receive: 0
            },
            startedDateTime: ''
        };
        entries.push(entry);
    }
    return entries;
}
exports.processEntries = processEntries;
function handleHeaders(headers) {
    if (!headers) {
        return [{ name: '', value: '' }];
    }
    return Object.entries(headers).map((e) => {
        return { name: e[0], value: e[1] };
    });
}
exports.handleHeaders = handleHeaders;
