// @tracerbench/core/dist/trace/export-hierarchy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const trace_1 = require("../trace");
function exportHierarchy(rawTraceData, hierarchy, trace, filePath, modMatcher) {
    const newTraceData = JSON.parse(JSON.stringify(rawTraceData));
    hierarchy.each(node => {
        const completeEvent = {
            pid: trace.mainProcess.id,
            tid: trace.mainProcess.mainThread.id,
            ts: node.data.min,
            ph: trace_1.TRACE_EVENT_PHASE_COMPLETE,
            cat: 'blink.user_timing',
            name: node.data.callFrame.functionName,
            args: {
                data: {
                    functionName: node.data.callFrame.functionName,
                    moduleName: modMatcher.findModuleName(node.data.callFrame)
                }
            },
            dur: node.data.max - node.data.min
        };
        newTraceData.traceEvents.push(completeEvent);
    });
    const outputFilePath = filePath.endsWith('.json')
        ? filePath.slice(0, filePath.length - 5)
        : filePath;
    fs.writeFileSync(`${outputFilePath}-processed.json`, JSON.stringify(newTraceData, null, ' '), 'utf8');
}
exports.exportHierarchy = exportHierarchy;
