// @tracerbench/core/dist/trace/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTab = exports.getNewTab = exports.createBrowser = exports.getBrowserArgs = exports.networkConditions = exports.liveTrace = void 0;
__exportStar(require("./archive-trace"), exports);
__exportStar(require("./auth"), exports);
var live_trace_1 = require("./live-trace");
Object.defineProperty(exports, "liveTrace", { enumerable: true, get: function () { return live_trace_1.liveTrace; } });
var conditions_1 = require("./conditions");
Object.defineProperty(exports, "networkConditions", { enumerable: true, get: function () { return conditions_1.networkConditions; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "getBrowserArgs", { enumerable: true, get: function () { return utils_1.getBrowserArgs; } });
Object.defineProperty(exports, "createBrowser", { enumerable: true, get: function () { return utils_1.createBrowser; } });
Object.defineProperty(exports, "getNewTab", { enumerable: true, get: function () { return utils_1.getNewTab; } });
Object.defineProperty(exports, "getTab", { enumerable: true, get: function () { return utils_1.getTab; } });
