// @tracerbench/core/dist/trace/trace-event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable import/export */
/* eslint-disable filenames/match-regex */
exports.TRACE_EVENT_PHASE_BEGIN = 'B';
exports.TRACE_EVENT_PHASE_END = 'E';
exports.TRACE_EVENT_PHASE_COMPLETE = 'X';
exports.TRACE_EVENT_PHASE_INSTANT = 'I';
exports.TRACE_EVENT_PHASE_ASYNC_BEGIN = 'S';
exports.TRACE_EVENT_PHASE_ASYNC_STEP_INTO = 'T';
exports.TRACE_EVENT_PHASE_ASYNC_STEP_PAST = 'p';
exports.TRACE_EVENT_PHASE_ASYNC_END = 'F';
exports.TRACE_EVENT_PHASE_NESTABLE_ASYNC_BEGIN = 'b';
exports.TRACE_EVENT_PHASE_NESTABLE_ASYNC_END = 'e';
exports.TRACE_EVENT_PHASE_NESTABLE_ASYNC_INSTANT = 'n';
exports.TRACE_EVENT_PHASE_FLOW_BEGIN = 's';
exports.TRACE_EVENT_PHASE_FLOW_STEP = 't';
exports.TRACE_EVENT_PHASE_FLOW_END = 'f';
exports.TRACE_EVENT_PHASE_METADATA = 'M';
exports.TRACE_EVENT_PHASE_COUNTER = 'C';
exports.TRACE_EVENT_PHASE_SAMPLE = 'P';
exports.TRACE_EVENT_PHASE_CREATE_OBJECT = 'N';
exports.TRACE_EVENT_PHASE_SNAPSHOT_OBJECT = 'O';
exports.TRACE_EVENT_PHASE_DELETE_OBJECT = 'D';
exports.TRACE_EVENT_PHASE_MEMORY_DUMP = 'v';
exports.TRACE_EVENT_PHASE_MARK = 'R';
exports.TRACE_EVENT_PHASE_CLOCK_SYNC = 'c';
exports.TRACE_EVENT_PHASE_ENTER_CONTEXT = '(';
exports.TRACE_EVENT_PHASE_LEAVE_CONTEXT = ')';
exports.TRACE_EVENT_PHASE_LINK_IDS = '=';
exports.TRACE_EVENT_SCOPE_NAME_GLOBAL = 'g';
exports.TRACE_EVENT_SCOPE_NAME_PROCESS = 'p';
exports.TRACE_EVENT_SCOPE_NAME_THREAD = 't';
