// @tracerbench/spawn-chrome/dist/defaultFlags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// reduce non determinism from background networking
exports.disableBackgroundNetworking = [
    // Disable translation
    "--disable-features=TranslateUI",
    // Disable several subsystems which run network requests in the background.
    "--disable-background-networking",
    // Enables the recording of metrics reports but disables reporting.
    "--metrics-recording-only",
    // Disable default component extensions with background pages
    "--disable-component-extensions-with-background-pages",
    // Disables syncing browser data to a Google Account.
    "--disable-sync",
    "--disable-client-side-phishing-detection",
    "--disable-component-update"
];
exports.disableTaskThrottling = [
    "--disable-renderer-backgrounding",
    "--disable-backgrounding-occluded-windows",
    "--disable-background-timer-throttling",
    "--disable-ipc-flooding-protection",
    "--disable-hang-monitor"
];
exports.disableFirstRun = [
    // Skip First Run tasks, whether or not it's actually the First Run.
    "--no-first-run",
    // Disables installation of default apps on first run.
    "--disable-default-apps",
    // Disables the default browser check.
    "--no-default-browser-check"
];
exports.automationFlags = [
    // Enable indication that browser is controlled by automation.
    "--enable-automation",
    // Unresponsive page dialog
    "--disable-hang-monitor",
    // Suppresses all error dialogs when present.
    "--noerrdialogs",
    // Prevents permission prompts from appearing by denying instead of showing
    // prompts.
    "--deny-permission-prompts",
    "--autoplay-policy=no-user-gesture-required",
    "--disable-popup-blocking",
    "--disable-prompt-on-repost",
    "--disable-search-geolocation-disclosure",
    // linux password store
    "--password-store=basic",
    // mac password store
    "--use-mock-keychain",
    "--force-color-profile=srgb"
];
exports.defaultFlags = exports.disableFirstRun.concat(exports.automationFlags, exports.disableTaskThrottling, exports.disableBackgroundNetworking);
exports.headlessFlags = [
    "--headless",
    "--hide-scrollbars",
    "--mute-audio"
];
exports.default = exports.defaultFlags;
