// @tracerbench/core/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var create_isolated_page_benchmark_1 = require("./create-isolated-page-benchmark");
Object.defineProperty(exports, "createIsolatedPageBenchmark", { enumerable: true, get: function () { return create_isolated_page_benchmark_1.default; } });
var create_trace_benchmark_1 = require("./create-trace-benchmark");
Object.defineProperty(exports, "createTraceBenchmark", { enumerable: true, get: function () { return create_trace_benchmark_1.default; } });
var create_trace_navigation_benchmark_1 = require("./create-trace-navigation-benchmark");
Object.defineProperty(exports, "createTraceNavigationBenchmark", { enumerable: true, get: function () { return create_trace_navigation_benchmark_1.default; } });
var run_1 = require("./run");
Object.defineProperty(exports, "run", { enumerable: true, get: function () { return run_1.default; } });
var trace_1 = require("./trace");
Object.defineProperty(exports, "liveTrace", { enumerable: true, get: function () { return trace_1.liveTrace; } });
Object.defineProperty(exports, "networkConditions", { enumerable: true, get: function () { return trace_1.networkConditions; } });
Object.defineProperty(exports, "getBrowserArgs", { enumerable: true, get: function () { return trace_1.getBrowserArgs; } });
Object.defineProperty(exports, "recordHARClient", { enumerable: true, get: function () { return trace_1.recordHARClient; } });
Object.defineProperty(exports, "authClient", { enumerable: true, get: function () { return trace_1.authClient; } });
