// tracerbench/dist/src/command-config/default-flag-args.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fidelityLookup = {
    test: 2,
    low: 20,
    medium: 30,
    high: 50,
};
// these default flag args are
// auto documented in the README.md
// in-addition to these we are defaulting
// chrome-debugging-client#defaultFlags.ts
exports.defaultFlagArgs = {
    plotTitle: 'TracerBench',
    cpuThrottleRate: 2,
    fidelity: 'low',
    markers: 'domComplete',
    browserArgs: [
        '--crash-dumps-dir=./tmp',
        '--disable-background-timer-throttling',
        '--disable-dev-shm-usage',
        '--disable-cache',
        '--disable-v8-idle-tasks',
        '--disable-breakpad',
        '--disable-notifications',
        '--disable-hang-monitor',
        '--safebrowsing-disable-auto-update',
        '--ignore-certificate-errors',
        '--v8-cache-options=none',
    ],
    methods: '""',
    tbResultsFolder: './tracerbench-results',
    iterations: 1,
    tracingLocationSearch: '?tracing',
    network: 'none',
    emulateDevice: '',
    emulateDeviceOrientation: 'vertical',
    dest: '',
    regressionThreshold: 100,
};
// specify with --headless flag
// ! --disable-gpu might cause issues with RHEL7
exports.headlessFlags = [
    '--headless',
    '--disable-gpu',
    '--hide-scrollbars',
    '--mute-audio',
    '--disable-logging',
];
function getDefaultValue(key) {
    if (key in exports.defaultFlagArgs) {
        return exports.defaultFlagArgs[key];
    }
}
exports.getDefaultValue = getDefaultValue;
