// @tracerbench/spawn/dist/newPipeMessageTransport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const newBufferSplitter_1 = require("./newBufferSplitter");
const newTaskQueue_1 = require("./newTaskQueue");
function newPipeMessageTransport(connect) {
    let attached = false;
    let closed = false;
    return (onMessage, onClose) => {
        if (attached) {
            throw new Error("already attached to transport");
        }
        attached = true;
        const [write, endWrite] = connect(onRead, onReadEnd, enqueueClose);
        const enqueue = newTaskQueue_1.default();
        const splitter = newBufferSplitter_1.default(0 /* NULL */, split => enqueueMessage(split.toString("utf8")));
        return sendMessage;
        function enqueueClose(error) {
            if (closed) {
                return;
            }
            closed = true;
            if (error) {
                enqueue(() => onClose(error));
            }
            else {
                enqueue(onClose);
            }
            endWrite();
        }
        function enqueueMessage(message) {
            enqueue(() => onMessage(message));
        }
        function onRead(data) {
            splitter.push(data);
        }
        function onReadEnd() {
            splitter.flush();
        }
        function sendMessage(message) {
            write(Buffer.from(message + "\0", "utf8"));
        }
    };
}
exports.default = newPipeMessageTransport;
