// @tracerbench/spawn/dist/newProcess.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const race_cancellation_1 = require("race-cancellation");
function newProcess(child, command, debugCallback) {
    let hasExited = false;
    let lastError;
    const emitter = new events_1.EventEmitter();
    const [raceExit, cancel] = race_cancellation_1.cancellableRace();
    child.on("error", error => {
        lastError = error;
        debugCallback("%o (pid: %o) 'error' event: %O", command, child.pid, error);
        onErrorOrExit(error);
    });
    child.on("exit", () => {
        debugCallback("%o (pid: %o) 'exit' event", command, child.pid);
        onErrorOrExit();
    });
    return {
        dispose,
        hasExited: () => hasExited,
        kill,
        off: emitter.removeListener.bind(emitter),
        on: emitter.on.bind(emitter),
        once: emitter.once.bind(emitter),
        raceExit,
        removeAllListeners: emitter.removeAllListeners.bind(emitter),
        removeListener: emitter.removeListener.bind(emitter),
        waitForExit,
    };
    /*
      https://nodejs.org/api/child_process.html#child_process_event_exit
  
      The 'exit' event may or may not fire after an error has occurred.
      When listening to both the 'exit' and 'error' events, it is important to guard against accidentally invoking handler functions multiple times.
     */
    function onErrorOrExit(error) {
        if (hasExited) {
            return;
        }
        hasExited = true;
        if (error) {
            cancel(`process exited early: ${error.message}`);
        }
        else {
            cancel(`process exited early`);
        }
        emitter.emit("exit");
    }
    async function exited(raceCancellation) {
        if (lastError) {
            throw lastError;
        }
        if (hasExited) {
            return;
        }
        return await race_cancellation_1.disposablePromise((resolve, reject) => {
            child.on("exit", resolve);
            child.on("error", reject);
            return () => {
                child.removeListener("exit", resolve);
                child.removeListener("error", reject);
            };
        }, raceCancellation);
    }
    async function waitForExit(timeout = 10000, raceCancellation) {
        if (hasExited) {
            return;
        }
        const result = await (timeout > 0
            ? race_cancellation_1.withRaceTimeout(exited, timeout)(raceCancellation)
            : exited(raceCancellation));
        return race_cancellation_1.throwIfCancelled(result);
    }
    async function kill(timeout, raceCancellation) {
        if (!child.killed && child.pid) {
            child.kill();
        }
        await waitForExit(timeout, raceCancellation);
    }
    async function dispose() {
        try {
            await kill();
        }
        catch (e) {
            // dispose is in finally and meant to be safe
            // we don't want to cover up error
            // just output for debugging
            debugCallback("%o (pid: %o) dispose error: %O", command, child.pid, e);
        }
    }
}
exports.default = newProcess;
